src/lib.rs

```diff
@@ -17,6 +18,228 @@ mod session;
+// session objects don't offer a way to store a `c_void` attached to a session.
+// So we use thread-locals. Before calling out to rustls code that may call
+// a callback, we set USERDATA for the current thread to the userdata pointer
+// for the current session. Before returning to the C caller, we restore
+// USERDATA to its previous value. Because a C callback may call back into
+// Rust code, we model these thread locals as a stack, so we can always
+// restore the previous version.
+thread_local! {
+    pub static USERDATA: RefCell<Vec<*mut c_void>> = RefCell::new(Vec::new());
+}
+
+/// UserdataGuard pops an entry off the USERDATA stack, restoring the
+/// thread-local state to its value previous to the creation of the UserdataGuard.
+/// Invariants: As long as a UserdataGuard is live:
+//// - The stack of userdata items for this thread must have at least one item.
+///  - The top item on that stack must be the one this guard was built with.
+///  - The `data` field must not be None.
+/// If any of these invariants fails, try_drop will return an error.
+pub struct UserdataGuard {
+    // Keep a copy of the data we expect to be popping off the stack. This allows
+    // us to check for consistency, and also serves to make this type !Send:
+    // https://doc.rust-lang.org/nightly/std/primitive.pointer.html#impl-Send-1
+    data: Option<*mut c_void>,
+}
+
+impl UserdataGuard {
+    fn new(u: *mut c_void) -> Self {
+        UserdataGuard { data: Some(u) }
+    }
+
+    /// Even though we have a Drop impl on this guard, when possible it's
+    /// best to call try_drop explicitly. That way any failures of internal
+    /// variants can be signaled to the user immediately by returning
+    /// rustls_result::Panic.
+    fn try_drop(mut self) -> Result<(), UserdataError> {
+        self.try_pop()
+    }
+
+    fn try_pop(&mut self) -> Result<(), UserdataError> {
+        let expected_data = match self.data {
+            Some(d) => d,
+            None => return Err(UserdataError::AlreadyPopped),
+        };
+        eprintln!("popping");
+        USERDATA
+            .try_with(|userdata| {
+                userdata.try_borrow_mut().map_or_else(
+                    |_| Err(UserdataError::AlreadyBorrowed),
+                    |mut v| match v.pop() {
+                        Some(u) => {
+                            self.data = None;
+                            if u == expected_data {
+                                Ok(())
+                            } else {
+                                Err(UserdataError::WrongData)
+                            }
+                        }
+                        None => Err(UserdataError::EmptyStack),
+                    },
+                )
+            })
+            .unwrap_or_else(|_: AccessError| Err(UserdataError::AccessError))
+    }
+}
+
+impl Drop for UserdataGuard {
+    fn drop(&mut self) {
+        self.try_pop().ok();
+    }
+}
+
+#[derive(Debug)]
```

Comment:
```suggestion
#[derive(Clone, Debug)]
```
Just because I think everything should be `Clone` if possible.